crates/proto/src/rr/rdata/tsig.rs

```diff
@@ -0,0 +1,770 @@
+    ///    | TSIG RDATA | Other Len      | in network byte order              |
+    ///    +------------+----------------+------------------------------------+
+    ///    | TSIG RDATA | Other Data     | exactly as transmitted             |
+    ///    +------------+----------------+------------------------------------+
+    /// ```
+    pub fn emit_tsig_for_mac(
+        &self,
+        encoder: &mut BinEncoder<'_>,
+        key_name: &Name,
+    ) -> ProtoResult<()> {
+        key_name.emit_as_canonical(encoder, true)?;
+        DNSClass::ANY.emit(encoder)?;
+        encoder.emit_u32(0)?; // TTL
+        self.algorithm.emit(encoder)?;
+        encoder.emit_u16((self.time >> 32) as u16)?;
+        encoder.emit_u32(self.time as u32)?;
+        encoder.emit_u16(self.fudge)?;
+        encoder.emit_u16(self.error)?;
+        encoder.emit_u16(self.other.len() as u16)?;
+        encoder.emit_vec(&self.other)?;
+        Ok(())
+    }
+
+    /// Add actual MAC value to existing TSIG record data.
+    ///
+    /// # Arguments
+    ///
+    /// * `mac` - mac to be stored in this record.
+    pub fn set_mac(self, mac: Vec<u8>) -> Self {
+        assert!(mac.len() < (1 << 16));
+        TSIG { mac, ..self }
+    }
+}
+
+/// Read the RData from the given Decoder
+///
+/// ```text
+///                       1 1 1 1 1 1 1 1 1 1 2 2 2 2 2 2 2 2 2 2 3 3
+///   0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1 2 3 4 5 6 7 8 9 0 1
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  /                         Algorithm Name                        /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                                                               |
+///  |          Time Signed          +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |                               |            Fudge              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          MAC Size             |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+             MAC               /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Original ID          |            Error              |
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+///  |          Other Len            |                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+           Other Data          /
+///  /                                                               /
+///  +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+/// ```
+pub fn read(decoder: &mut BinDecoder<'_>) -> ProtoResult<TSIG> {
+    let algorithm = Algorithm::read(decoder)?;
+    let time_high = decoder.read_u16()?.unverified(/*valid as any u16*/) as u64;
+    let time_low = decoder.read_u32()?.unverified(/*valid as any u32*/) as u64;
+    let time = (time_high << 32) + time_low;
```

Comment:
any concerns with overflow here? we may want to use a checked add. Also, maybe zero out the high-order bits on time_low (I know it's pedantic as that should happen with the `u16` cast to `u64`)?

Reply:
The addition can't overflow, adding a number on 48 bits (u16 << 32) and a number on 32 bits uses at most 49 bits. In this case as they are disjoints (no bit in common), the result will actually be at most 48 bits too. In fact instead of summing, they could be or-ed (or xor-ed). I should probably do that to make it clear